// @microsoft/power-apps-actions/dist/Common/parameterPropertyInfo.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
/*
 * Parameter Info class for storing parameter property information
 */
export class ParameterPropertyInfo {
    propertyName;
    propertyDefinition;
    originalDefinition;
    rawParamName;
    isRequired;
    /**
     * Constructs a new ParameterPropertyInfo instance.
     * @param propertyName The sanitized identifier name of the property (safe for use in TS function signatures).
     * @param propertyDefinition The property definition using the sanitized name (e.g. `deployment_id?: string`).
     * @param originalDefinition The property definition using the original (quoted) name for object type literals
     *   (e.g. `"deployment-id"?: string`). Defaults to `propertyDefinition` when the name needs no quoting.
     * @param rawParamName The original (unsanitized) API parameter name (e.g. `"azureml-model-deployment"`).
     *   Used to emit the correct mapping key in params object literals. Defaults to `propertyName`.
     * @param isRequired Whether this parameter is required (i.e. non-optional in the generated type).
     *   Stored as structured data to avoid brittle string-matching on `propertyDefinition`.
     */
    constructor(propertyName, propertyDefinition, originalDefinition = propertyDefinition, rawParamName = propertyName, isRequired = false) {
        this.propertyName = propertyName;
        this.propertyDefinition = propertyDefinition;
        this.originalDefinition = originalDefinition;
        this.rawParamName = rawParamName;
        this.isRequired = isRequired;
    }
}
