// @microsoft/power-apps-actions/dist/Actions/ListFlows.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { DataverseSystemDataService, getEnvironmentByName, getPlayerServiceConfig, } from '../services/index.js';
import { SUPPORTED_TRIGGER_KINDS } from './flowConstants.js';
/**
 * Determines whether a flow is supported by Code Apps by inspecting its trigger kind.
 * Only flows with a PowerApp or PowerAppV2 manual trigger are supported.
 *
 * @param dataverseWorkflow The Dataverse workflow to check
 * @returns true if the flow has a supported trigger kind
 */
export function isSupportedCodeAppFlow(dataverseWorkflow) {
    try {
        const clientData = JSON.parse(dataverseWorkflow.clientdata || '{}');
        const triggers = clientData?.properties?.definition?.triggers;
        if (!triggers)
            return false;
        return Object.values(triggers).some((trigger) => trigger?.kind && SUPPORTED_TRIGGER_KINDS.has(trigger.kind));
    }
    catch {
        return false;
    }
}
/**
 * Maps a DataverseWorkflow to the public Workflow interface
 * @param dataverseWorkflow The internal Dataverse workflow object
 * @returns A Workflow object with cleaned up field names
 */
function mapToWorkflow(dataverseWorkflow) {
    return {
        name: dataverseWorkflow.name,
        modifiedOn: dataverseWorkflow.modifiedon,
        workflowId: dataverseWorkflow.resourceid,
        statecode: dataverseWorkflow.statecode,
    };
}
/**
 * Get all cloud flows from a Dataverse environment that are supported by Code Apps.
 * Only flows with a PowerApp or PowerAppV2 trigger are returned.
 * The org URL is derived from the environment ID in the context.
 *
 * @param {ListFlowsContext} context
 * @returns {Promise<Workflow[]>}
 */
export async function listFlowsAsync(context) {
    // Get environment metadata to retrieve the Dataverse org URL
    const metadata = await getEnvironmentByName(context.environmentName);
    const orgUrl = metadata?.properties?.linkedEnvironmentMetadata?.instanceApiUrl;
    if (!orgUrl) {
        throw new Error('Unable to determine the Dataverse organization URL. Ensure the environment has a linked Dataverse instance.');
    }
    const { httpClient } = getPlayerServiceConfig();
    const dataService = new DataverseSystemDataService(httpClient);
    const dataverseWorkflows = (await dataService.getWorkflows(orgUrl, context.logger, context.actionsParams.search)) || [];
    return dataverseWorkflows.filter(isSupportedCodeAppFlow).map(mapToWorkflow);
}
